File: services/go-core/internal/plugins/domain/entities.go
package domain

import (
	"context"
	"time"

	"github.com/google/uuid"
)

// PluginCapability representa el tipo de IA
type PluginCapability string

const (
	CapabilityStemSeparation  PluginCapability = "stem-separation"
	CapabilityEmbeddings      PluginCapability = "embeddings"
	CapabilityKnowledge       PluginCapability = "knowledge"
	CapabilityRecommendations PluginCapability = "recommendations"
	CapabilityDownload        PluginCapability = "download"
)

// JobStatus representa el estado de un trabajo en el plugin
type JobStatus string

const (
	JobPending    JobStatus = "pending"
	JobProcessing JobStatus = "processing"
	JobCompleted  JobStatus = "completed"
	JobFailed     JobStatus = "failed"
	JobCancelled  JobStatus = "cancelled"
)

// Manifest representa la identidad de un plugin
type Manifest struct {
	ID           string            `json:"id"`
	Name         string            `json:"name"`
	Version      string            `json:"version"`
	Capability   PluginCapability  `json:"capability"`
	Description  string            `json:"description"`
	Model        string            `json:"model"`
	Requirements map[string]string `json:"requirements"`
}

// HealthStatus representa el estado de salud del plugin
type HealthStatus struct {
	Status       string    `json:"status"`
	Timestamp    time.Time `json:"timestamp"`
	GPUAvailable bool      `json:"gpu_available"`
	ActiveJobs   int       `json:"active_jobs"`
	ModelCached  bool      `json:"model_cached"`
	StorageUsage int64     `json:"storage_usage_bytes"` // New field
}

// JobPriority representa la prioridad de un trabajo
type JobPriority int

const (
	PriorityStreaming JobPriority = 0
	PriorityNormal    JobPriority = 10
	PriorityLow       JobPriority = 20
)

// JobResponse representa la respuesta de un trabajo
type JobResponse struct {
	ID        string            `json:"id"`
	TrackID   string            `json:"track_id"`
	Status    JobStatus         `json:"status"`
	Priority  JobPriority       `json:"priority"`
	Progress  float64           `json:"progress"`
	Result    map[string]string `json:"result,omitempty"`
	Error     string            `json:"error,omitempty"`
	CreatedAt time.Time         `json:"created_at"`
	UpdatedAt time.Time         `json:"updated_at"`
}

// AIPlugin representa un plugin registrado en el sistema
type AIPlugin struct {
	ID        uuid.UUID              `json:"id"`
	Manifest  Manifest               `json:"manifest"`
	BaseURL   string                 `json:"baseUrl"`
	Health    HealthStatus           `json:"health"`
	IsEnabled bool                   `json:"isEnabled"`
	Config    map[string]interface{} `json:"config"`
}

// RecommendationRequest represents a request for recommendations
type RecommendationRequest struct {
	TrackID   string   `json:"track_id,omitempty"`
	ArtistID  string   `json:"artist_id,omitempty"`
	Limit     int      `json:"limit"`
	Context   []string `json:"context,omitempty"` // e.g. "favorites", "recent"
	Diversity float64  `json:"diversity"`         // 0.0 (Similar) - 1.0 (Diverse)
}

// Recommendation represents a recommended item
type Recommendation struct {
	ID     string  `json:"id"`
	Type   string  `json:"type"` // track, artist, album
	Score  float64 `json:"score"`
	Reason string  `json:"reason"`
}

// IPluginClient define la interfaz para comunicarse con cualquier plugin AI (Port)
type IPluginClient interface {
	GetManifest(ctx context.Context, baseURL string) (*Manifest, error)
	GetHealth(ctx context.Context, baseURL string) (*HealthStatus, error)
	CreateJob(ctx context.Context, baseURL string, trackID, filePath string, priority JobPriority, stems []string) (*JobResponse, error)
	GetJobStatus(ctx context.Context, baseURL string, jobID string) (*JobResponse, error)
	CancelJob(ctx context.Context, baseURL string, jobID string) error
	GetRecommendations(ctx context.Context, baseURL string, req RecommendationRequest) ([]Recommendation, error)
	UpdateConfig(ctx context.Context, baseURL string, config map[string]interface{}) error
	DeleteData(ctx context.Context, baseURL string) error
}

File: services/ai-plugins/brain/src/domain/entities.py
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import List, Optional, Dict, Any
import uuid

class JobStatus(str, Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"

class JobPriority(int, Enum):
    """Job priority levels (Lower value = Higher priority)"""
    STREAMING = 0   # Immediate priority
    NORMAL = 10     # Standard
    LOW = 20        # Background

class CapabilityType(str, Enum):
    EMBEDDINGS = "embeddings"

@dataclass(frozen=True)
class PluginCapability:
    type: CapabilityType
    name: str
    version: str
    description: str

@dataclass(frozen=True)
class EmbeddingJob:
    id: str
    track_id: str
    file_path: str
    status: JobStatus = JobStatus.PENDING
    priority: JobPriority = JobPriority.NORMAL
    progress: float = 0.0
    embedding: Optional[List[float]] = None
    model_version: Optional[str] = None
    error: Optional[str] = None
    created_at: datetime = field(default_factory=datetime.now)
    updated_at: datetime = field(default_factory=datetime.now)

    @property
    def is_terminal(self) -> bool:
        return self.status in [JobStatus.COMPLETED, JobStatus.FAILED, JobStatus.CANCELLED]

    def mark_processing(self) -> 'EmbeddingJob':
        return EmbeddingJob(**{**self.__dict__, "status": JobStatus.PROCESSING, "updated_at": datetime.now()})

    def mark_completed(self, embedding: List[float], model_version: str) -> 'EmbeddingJob':
        return EmbeddingJob(**{
            **self.__dict__,
            "status": JobStatus.COMPLETED,
            "progress": 100.0,
            "embedding": embedding,
            "model_version": model_version,
            "updated_at": datetime.now()
        })

    def mark_failed(self, error: str) -> 'EmbeddingJob':
        return EmbeddingJob(**{
            **self.__dict__,
            "status": JobStatus.FAILED,
            "error": error,
            "updated_at": datetime.now()
        })

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "track_id": self.track_id,
            "file_path": self.file_path,
            "status": self.status.value,
            "priority": int(self.priority),
            "progress": self.progress,
            "embedding": self.embedding,
            "model_version": self.model_version,
            "error": self.error,
            "created_at": self.created_at.strftime('%Y-%m-%dT%H:%M:%SZ'),
            "updated_at": self.updated_at.strftime('%Y-%m-%dT%H:%M:%SZ')
        }

@dataclass(frozen=True)
class HealthStatus:
    status: str
    version: str
    uptime_seconds: float
    gpu_available: bool
    gpu_name: Optional[str] = None
    memory_usage_mb: float = 0.0
    active_jobs: int = 0

@dataclass(frozen=True)
class Recommendation:
    id: str
    type: str # track, artist, album
    score: float
    reason: str
